// ngx-easy-table/esm5/lib/pipes/sort.pipe.js
import { __decorate } from "tslib";
import { Pipe } from '@angular/core';
import { FiltersService } from '../services/filters.service';
var SortPipe = /** @class */ (function () {
    function SortPipe() {
        this.defaultArray = [];
    }
    SortPipe_1 = SortPipe;
    SortPipe.isNaN = function (aV, bV) {
        return (isNaN(parseFloat(aV)) || !isFinite(aV)) || (isNaN(parseFloat(bV)) || !isFinite(bV));
    };
    SortPipe.compare = function (a, b, key) {
        var split = key.split('.');
        var aPath = FiltersService.getPath(split, a);
        var bPath = FiltersService.getPath(split, b);
        var aValue = (aPath + '').toLowerCase();
        var bValue = (bPath + '').toLowerCase();
        if (SortPipe_1.isNaN(aPath, bPath)) {
            return aValue.localeCompare(bValue);
        }
        else {
            if (parseFloat(aPath) < parseFloat(bPath)) {
                return -1;
            }
            if (parseFloat(aPath) > parseFloat(bPath)) {
                return 1;
            }
        }
        return 0;
    };
    SortPipe.prototype.transform = function (array, filter, config) {
        if (this.defaultArray.length === 0) {
            this.defaultArray = array;
        }
        if (!filter.key || filter.key === '') {
            return array;
        }
        if (filter.order === '') {
            return this.defaultArray;
        }
        if (filter.order === 'asc') {
            return this.sortAsc(array, filter, config);
        }
        else {
            return this.sortDesc(array, filter, config);
        }
    };
    SortPipe.prototype.sortAsc = function (array, filter, config) {
        if (config && config.groupRows) {
            return array.map(function (arr) { return arr.sort(function (a, b) { return SortPipe_1.compare(a, b, filter.key); }); });
        }
        return array.sort(function (a, b) { return SortPipe_1.compare(a, b, filter.key); });
    };
    SortPipe.prototype.sortDesc = function (array, filter, config) {
        if (config && config.groupRows) {
            return array.map(function (arr) { return arr.sort(function (a, b) { return SortPipe_1.compare(b, a, filter.key); }); });
        }
        return array.sort(function (a, b) { return SortPipe_1.compare(b, a, filter.key); });
    };
    var SortPipe_1;
    SortPipe = SortPipe_1 = __decorate([
        Pipe({
            name: 'sort',
        })
    ], SortPipe);
    return SortPipe;
}());
export { SortPipe };
