// ngx-easy-table/esm5/lib/services/config-service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { STYLE, THEME } from '../model/config';
export var DefaultConfig = {
    searchEnabled: false,
    headerEnabled: true,
    orderEnabled: true,
    orderEventOnly: false,
    paginationEnabled: true,
    exportEnabled: false,
    clickEvent: true,
    selectRow: false,
    selectCol: false,
    selectCell: false,
    rows: 10,
    additionalActions: false,
    serverPagination: false,
    isLoading: false,
    detailsTemplate: false,
    groupRows: false,
    paginationRangeEnabled: true,
    collapseAllRows: false,
    checkboxes: false,
    radio: false,
    resizeColumn: false,
    fixedColumnWidth: true,
    horizontalScroll: false,
    draggable: false,
    logger: false,
    showDetailsArrow: false,
    showContextMenu: false,
    persistState: false,
    paginationMaxSize: 5,
    threeWaySort: false,
    tableLayout: {
        style: STYLE.NORMAL,
        theme: THEME.LIGHT,
        borderless: false,
        hover: true,
        striped: false,
    },
};
var DefaultConfigService = /** @class */ (function () {
    function DefaultConfigService() {
    }
    DefaultConfigService.config = DefaultConfig;
    DefaultConfigService = __decorate([
        Injectable()
    ], DefaultConfigService);
    return DefaultConfigService;
}());
export { DefaultConfigService };
