// ngx-easy-table/esm5/lib/services/style.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
/* tslint:disable:no-useless-cast */
var StyleService = /** @class */ (function () {
    function StyleService() {
    }
    StyleService.prototype.setRowClass = function (val) {
        var selector = "#table > tbody > tr:nth-child(" + val.row + ")";
        var row = document.querySelector(selector);
        if (row) {
            row.classList.add(val.className);
        }
    };
    StyleService.prototype.setCellClass = function (val) {
        var selector = "#table > tbody > tr:nth-child(" + val.row + ") > td:nth-child(" + val.cell + ")";
        var cell = document.querySelector(selector);
        if (cell) {
            cell.classList.add(val.className);
        }
    };
    StyleService.prototype.setRowStyle = function (val) {
        var selector = "#table > tbody > tr:nth-child(" + val.row + ")";
        var row = document.querySelector(selector);
        if (row) {
            // tslint:disable-next-line:no-string-literal
            row.style[val.attr] = val.value;
        }
    };
    StyleService.prototype.setCellStyle = function (val) {
        var selector = "#table > tbody > tr:nth-child(" + val.row + ") > td:nth-child(" + val.cell + ")";
        var cell = document.querySelector(selector);
        if (cell) {
            // tslint:disable-next-line:no-string-literal
            cell.style[val.attr] = val.value;
        }
    };
    StyleService.prototype.pinnedWidth = function (pinned, column) {
        if (pinned) {
            return 150 * column + 'px';
        }
    };
    StyleService = __decorate([
        Injectable()
    ], StyleService);
    return StyleService;
}());
export { StyleService };
