// ngx-easy-table/esm2015/lib/components/header/header.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
let HeaderComponent = class HeaderComponent {
    constructor() {
        this.update = new EventEmitter();
    }
    unifyKey(key) {
        return key.replace('.', '_');
    }
    onSearch(input) {
        this.update.emit([
            { value: input.value, key: this.column.key },
        ]);
    }
};
__decorate([
    Input(),
    __metadata("design:type", Object)
], HeaderComponent.prototype, "column", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], HeaderComponent.prototype, "update", void 0);
HeaderComponent = __decorate([
    Component({
        selector: 'table-header',
        template: "<label for=\"search_{{ unifyKey(column.key) }}\">\n  <input type=\"text\"\n         id=\"search_{{ unifyKey(column.key) }}\"\n         aria-label=\"Search\"\n         placeholder=\"{{ column.placeholder ? column.placeholder : column.title }}\"\n         class=\"ngx-table__header-search\"\n         #input\n         (input)=\"onSearch(input)\"\n  >\n</label>\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], HeaderComponent);
export { HeaderComponent };
